src/backend/distributed/executor/multi_router_executor.c

```diff
@@ -620,3 +756,109 @@ RouterExecutorEnd(QueryDesc *queryDesc)
 	queryDesc->estate = NULL;
 	queryDesc->totaltime = NULL;
 }
+
+
+/*
+ * RouterTransactionCallback handles committing or aborting remote transactions
+ * after the local one has committed or aborted. It only sends COMMIT or ABORT
+ * commands to still-healthy remotes; the failed ones are marked as inactive if
+ * after a successful COMMIT (no need to mark on ABORTs).
+ */
+static void
+RouterTransactionCallback(XactEvent event, void *arg)
+{
+	if (!txnStarted)
+	{
+		return;
+	}
+
+	switch (event)
+	{
+		/* after the local transaction commits, commit the remote ones */
+		case XACT_EVENT_COMMIT:
+#if (PG_VERSION_NUM >= 90500)
+		case XACT_EVENT_PARALLEL_COMMIT:
+#endif
+		{
+			ListCell *failedPlacementCell = NULL;
+
+			ExecuteTransactionEnd("COMMIT TRANSACTION");
+
+			/* now mark all failed placements inactive */
+			foreach(failedPlacementCell, failedParticipantList)
+			{
+				ShardPlacement *failedPlacement = NULL;
+				failedPlacement = (ShardPlacement *) lfirst(failedPlacementCell);
+				uint64 shardLength = 0;
+
+				DeleteShardPlacementRow(failedPlacement->shardId,
+										failedPlacement->nodeName,
+										failedPlacement->nodePort);
+				InsertShardPlacementRow(failedPlacement->shardId,
+										FILE_INACTIVE, shardLength,
+										failedPlacement->nodeName,
+										failedPlacement->nodePort);
+			}
+
+			break;
+		}
+
+		/* if the local transaction aborts, send ABORT to healthy remotes */
+		case XACT_EVENT_ABORT:
+#if (PG_VERSION_NUM >= 90500)
+		case XACT_EVENT_PARALLEL_ABORT:
+#endif
+		{
+			ExecuteTransactionEnd("ABORT TRANSACTION");
+
+			break;
+		}
+
+		/* no support for prepare with long-running transactions */
+		case XACT_EVENT_PREPARE:
+		case XACT_EVENT_PRE_PREPARE:
+		{
+			ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+							errmsg("cannot prepare a transaction that modified "
+								   "distributed tables")));
+
+			break;
+		}
+
+		/* we don't have any work to do pre-commit */
+		case XACT_EVENT_PRE_COMMIT:
+#if (PG_VERSION_NUM >= 90500)
+		case XACT_EVENT_PARALLEL_PRE_COMMIT:
+#endif
+		{
+			/* leave early to avoid resetting transaction state */
+			return;
+		}
+	}
+
+	/* reset transaction state */
+	shardForTransaction = INVALID_SHARD_ID;
+	participantList = NIL;
+	failedParticipantList = NIL;
+	txnStarted = false;
+}
+
+
+/*
+ * pgfdw_subxact_callback --- cleanup at subtransaction end.
+ */
+static void
+RouterSubtransactionCallback(SubXactEvent event, SubTransactionId subId,
+							 SubTransactionId parentSubid, void *arg)
+{
+	if (txnStarted && event == SUBXACT_EVENT_ABORT_SUB)
+	{
+		UnregisterSubXactCallback(RouterSubtransactionCallback, NULL);
+
+		AbortOutOfAnyTransaction();
```

Comment:
Adding another `static` file variable to remember that a subxact abort was attempted, then erroring in the main xact callback.

Still kind of weird, since the user will be able to deduce that the `ROLLBACK` wasn't actually applied using e.g. `SELECT`s within the remaining transaction. No biggie since it won't be applied anyway.